File: backend/src/controllers/wardrobeController.js
/**
 * Wardrobe Controller
 * Handles HTTP requests for wardrobe management
 */

const wardrobeService = require('../services/wardrobeService');

class WardrobeController {
  /**
   * GET /api/wardrobe
   * Get all wardrobe items
   */
  async getAllItems(req, res) {
    try {
      const items = wardrobeService.getAllItems();
      
      res.json({
        success: true,
        data: items,
        count: items.length
      });
    } catch (error) {
      console.error('Error in getAllItems:', error);
      res.status(500).json({
        error: {
          message: 'Failed to retrieve wardrobe items'
        }
      });
    }
  }

  /**
   * POST /api/wardrobe
   * Add new item to wardrobe
   */
  async addItem(req, res) {
    try {
      const { analysis, imageData } = req.body;

      // Validation
      if (!analysis) {
        return res.status(400).json({
          error: { message: 'Analysis data is required' }
        });
      }

      // Add item via service
      const newItem = wardrobeService.addItem(analysis, imageData);

      res.status(201).json({
        success: true,
        data: newItem
      });
    } catch (error) {
      console.error('Error in addItem:', error);
      res.status(500).json({
        error: {
          message: 'Failed to add item to wardrobe'
        }
      });
    }
  }

  /**
   * DELETE /api/wardrobe/:id
   * Remove item from wardrobe
   */
  async deleteItem(req, res) {
    try {
      const { id } = req.params;

      const deleted = wardrobeService.deleteItem(id);

      if (!deleted) {
        return res.status(404).json({
          error: { message: 'Item not found' }
        });
      }

      res.json({
        success: true,
        message: 'Item removed from wardrobe'
      });
    } catch (error) {
      console.error('Error in deleteItem:', error);
      res.status(500).json({
        error: {
          message: 'Failed to delete item'
        }
      });
    }
  }

  /**
   * PATCH /api/wardrobe/:id/favorite
   * Toggle favorite status
   */
  async toggleFavorite(req, res) {
    try {
      const { id } = req.params;

      const item = wardrobeService.toggleFavorite(id);

      if (!item) {
        return res.status(404).json({
          error: { message: 'Item not found' }
        });
      }

      res.json({
        success: true,
        data: item
      });
    } catch (error) {
      console.error('Error in toggleFavorite:', error);
      res.status(500).json({
        error: {
          message: 'Failed to update favorite status'
        }
      });
    }
  }

  /**
   * DELETE /api/wardrobe
   * Clear entire wardrobe
   */
  async clearWardrobe(req, res) {
    try {
      const count = wardrobeService.clearWardrobe();

      res.json({
        success: true,
        message: 'Wardrobe cleared',
        itemsCleared: count
      });
    } catch (error) {
      console.error('Error in clearWardrobe:', error);
      res.status(500).json({
        error: {
          message: 'Failed to clear wardrobe'
        }
      });
    }
  }

  /**
   * GET /api/wardrobe/statistics
   * Get wardrobe statistics
   */
  async getStatistics(req, res) {
    try {
      const stats = wardrobeService.getStatistics();

      res.json({
        success: true,
        data: stats
      });
    } catch (error) {
      console.error('Error in getStatistics:', error);
      res.status(500).json({
        error: {
          message: 'Failed to get statistics'
        }
      });
    }
  }
}

module.exports = new WardrobeController();

File: backend-python/controllers/wardrobe_controller.py
"""
Wardrobe Controller
Handles HTTP requests for wardrobe management
"""
from flask import request, jsonify
from services.wardrobe_service import wardrobe_service

class WardrobeController:
    """Controller for wardrobe endpoints"""
    
    def get_all_items(self):
        """
        GET /api/wardrobe
        Get all wardrobe items
        """
        try:
            items = wardrobe_service.get_all_items()
            return jsonify({
                'success': True,
                'data': items
            }), 200
        except Exception as e:
            print(f'Error in getAllItems controller: {str(e)}')
            return jsonify({
                'success': False,
                'error': 'Failed to fetch items'
            }), 500
    
    def add_item(self):
        """
        POST /api/wardrobe
        Add item to wardrobe (typically done via analyze endpoint)
        """
        try:
            data = request.get_json()
            
            # Support both formats: from analyze endpoint and from frontend
            if not data or 'analysis' not in data:
                return jsonify({
                    'success': False,
                    'error': 'Missing required fields'
                }), 400
            
            # Get imageInfo - either from data or create from imageData
            image_info = data.get('imageInfo')
            if not image_info and 'imageData' in data:
                # Create basic imageInfo from imageData
                image_info = {
                    'filename': 'uploaded_image.jpg',
                    'mimetype': 'image/jpeg',
                    'data': data['imageData']
                }
            
            if not image_info:
                return jsonify({
                    'success': False,
                    'error': 'Missing image information'
                }), 400
            
            item = wardrobe_service.add_item(image_info, data['analysis'])
            
            return jsonify({
                'success': True,
                'data': item
            }), 201
            
        except Exception as e:
            print(f'Error in addItem controller: {str(e)}')
            import traceback
            traceback.print_exc()
            return jsonify({
                'success': False,
                'error': 'Failed to add item'
            }), 500
    
    def delete_item(self, item_id: int):
        """
        DELETE /api/wardrobe/:id
        Delete item from wardrobe
        """
        try:
            success = wardrobe_service.delete_item(item_id)
            
            if not success:
                return jsonify({
                    'success': False,
                    'error': 'Item not found'
                }), 404
            
            return jsonify({
                'success': True,
                'message': 'Item deleted successfully'
            }), 200
            
        except Exception as e:
            print(f'Error in deleteItem controller: {str(e)}')
            return jsonify({
                'success': False,
                'error': 'Failed to delete item'
            }), 500
    
    def toggle_favorite(self, item_id: int):
        """
        PATCH /api/wardrobe/:id/favorite
        Toggle favorite status
        """
        try:
            item = wardrobe_service.toggle_favorite(item_id)
            
            if not item:
                return jsonify({
                    'success': False,
                    'error': 'Item not found'
                }), 404
            
            return jsonify({
                'success': True,
                'data': item
            }), 200
            
        except Exception as e:
            print(f'Error in toggleFavorite controller: {str(e)}')
            return jsonify({
                'success': False,
                'error': 'Failed to update favorite status'
            }), 500
    
    def clear_wardrobe(self):
        """
        DELETE /api/wardrobe
        Clear all items from wardrobe
        """
        try:
            wardrobe_service.clear_wardrobe()
            
            return jsonify({
                'success': True,
                'message': 'Wardrobe cleared successfully'
            }), 200
            
        except Exception as e:
            print(f'Error in clearWardrobe controller: {str(e)}')
            return jsonify({
                'success': False,
                'error': 'Failed to clear wardrobe'
            }), 500
    
    def get_statistics(self):
        """
        GET /api/wardrobe/statistics
        Get wardrobe statistics
        """
        try:
            stats = wardrobe_service.get_statistics()
            
            return jsonify({
                'success': True,
                'data': stats
            }), 200
            
        except Exception as e:
            print(f'Error in getStatistics controller: {str(e)}')
            return jsonify({
                'success': False,
                'error': 'Failed to fetch statistics'
            }), 500

# Create singleton instance
wardrobe_controller = WardrobeController()
